### ISSUE TITLE
update feed scraping

### ISSUE THREAD

**Issue by hroberts** [berkmancenter/mediacloud#227]
Our current feed scraping code hasn't been touched much for the past couple of years, and I think it is not nearly as accurate as it once was.  We need to update the code to make sure we're doing the best job possible of finding a feed for each source if it exists.

As with the date guessing stuff, we need to run this as a data driven task, so it should start with creating a training / evaluation set of media sources and their feeds and validating whatever solutions you come up with.  That process should also include validating our current code so that we know what the baseline is.  It would be great to have at least evaluation sets -- one for U.S. sites and one for international sites (the abyz sets would be terrific).

Some of the work of making those training sets will be just manually poking at sites to find rss feeds, but you could also rely on existing found feeds or feeds found from the new code to identify the feeds that you will evaluate (so if our code finds a feed for a given source, you could just manually validate that feed rather than doing the whole feed discovery process manually).  FWIW, I find that searching for 'rss nytimes.com' (obv replacing the url) is by far the fastest way to find rss for a given site manually (if we could figure out a way to automate that without getting blocked by google it would be a great discovery mechanism!).

The current code is here:

https://github.com/berkmancenter/mediacloud/blob/master/lib/MediaWords/Feed/Scrape.pm

The main place this code is called from is our media rescraping job:

https://github.com/berkmancenter/mediacloud/blob/master/lib/MediaWords/DBI/Media/Rescrape.pm

That rescraping code runs both when we initially add a source to the system and also every six months.  When it is run for a site that already has feeds, it will keep any feeds the source already has but add any new ones discovered by the new feed scraping.

The basic idea of the Feed::Scrape code is to download the url of the media source, look for any links in the html that have a url or link text that look like they might be rss, and download each one of those and try to parse it as rss/rdf/atom.  Then for any link whose content looks like html instead of rss/rdf/atom, repeat the above process (but only recurse this way for one level).  This catches the case that many sites have of putting a link to an rss index page on the main page rather than a link to rss itself.

The code also has some shortcuts to try a number of different common rss locations (eg. $url/feed/rss) for all sites and some special cases for special blog hosting sites (we know where the feed should be for all blogger sites, for example).

In addition to the basic task of finding all of the feeds for a given site, we have to weed out sites that for whatever reason include the links to the rss of others sites on their sites.  This comes up in some cases for sites owned by companies with multiple publications, but it also just happens radomly, as some people link to the rss feeds of other sites that they like or whatever.  

To lessen the chance of this happening, we try to identify what we call 'default feeds'.  If we find a default feed, we just ignore everything else and assume that we can capture all of the content of the site from that default feed (for example, if we find http://nytimes.com/feed has valid rss, we don't bother to include any other feeds).  This default feed identification also lets us track just one feed for a given site rather than potentially hundreds if we can't identify some single default feed.

All of this is a big heavy hammer to make it as likely as possible that we will find a feed if one exists.  We are much, much more concerned about maximizing recall for the process than about its performance.  If it takes us an hour of downloading a bunch of cruft to discover a feed for a given source, it is easily worth the processing time.

One source of large potential improvement is to use the feedly api to search for possible rss feeds.  Last time I checked feedly provided this functionality in their free, un-authenticated api.  And I commonly find in feedly rss feeds that are neither published anywhere I can see on a given site or found by the 'rss nytimes.com' google search.  I assume that's because feedly is basically a crowd sourced discovery platform -- all it requires is one person to find the rss feed, and then it is in the sytem for everyone to find.

**Comment by pypt** [comment]
https://github.com/berkmancenter/mediacloud/pull/222 fixes this, right?

Colin, do you have stats for how much better / worse your PR is doing at finding feeds?

**Comment by ColCarroll** [comment]
I do not – happy to run this on a list of urls/story_ids if you can run the perl script on those.

**Comment by hroberts** [comment]
We need to create an evaluation set for this so that we can judge its precision and recall, both against the existing code and against variations of the heuristics in the new code.

For the evaluation set, we should pull maybe 50 random sources from each of:

* US Top Online Media
* Retweet Partisanship left/right
* 4 countries from the recently imported abyz sets

For each of those sources, we need to do the best we can do to create a ground truth of what the rss feed or feeds for each source are, then use that ground truth for evaluation. It is important that the programmer not directly compare the individual results in the evaluation set to the output of the program, because that will end up overfitting the algorithm and overstating the accuracy.  Some amount of overfitting is unavoidable if the programmer is the one creating the evaluation set in the first place (you'll just remember the kinds of places you found feeds), but we should at least make sure we are directly optimizing on the evaluation set

I usually find it helpful when doing this sort of manual heuristic evaluation to create a second, maybe smaller, training set that I can use to for optimization (looking for specific cases where the algorithm failed), which is fine as long as we are using the evaluation set for the ultimate accuracy numbers.

**Comment by rahulbot** [comment]
This is a hard thing to validate, so lets do this one step at a time.
1. Colin can you generate a list of 50 total random media, sourced from the collections Hal mentions above?
2. I'll see if I can get the research team to fill in a spreadsheet of "url" and "rss_url" from their manual scraping process. This will be our ground-truth to compare against, so it is best if someone else does it.
3. Colin will compare those to assess:
  a) precision and recall when it comes to just the RSS urls discovered
  b) comparison of stories contained in RSS feeds to see if the machine-discovered ones are a super-set of manually-discovered ones

I hope I've described this cogently.  Ground truth is hard to know here, so lets do our best to validate a hard thing.

**Comment by ColCarroll** [comment]
Hi @hroberts I am not sure how to find urls from "US Top Online Media" or "Retweet Partisanship left/right".  Also, you suggested getting 50 urls from each of a few sources (totaling either 150 or 250 urls), and @rahulbot suggested 50 total random urls. 

In case it helps, I grabbed 250 urls randomly from the complete set of media cloud urls, which [can be found here](https://gist.github.com/ColCarroll/6d3533872f6ef29e8f56e0967aa1f170).

**Comment by rahulbot** [comment]
I've asked the ML research team if they have someone that can do the manual data generation for us.

**Comment by rahulbot** [comment]
Update: I wrote a [quick script](https://gist.github.com/rahulbot/6371c35677655305d2988e77f7b3af05) to generate a list of 50 random sources.  I've sent this to Anissa to manually pull RSS feeds: [random-sources.txt](https://github.com/berkmancenter/mediacloud/files/1803994/random-sources.txt)

**Comment by rahulbot** [comment]
Integrating this is blocking on the validation work.
